docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+that is configured as follows:
+
+```go
+AffinitySelector {
+	HardAffinityExpressions: {{PodSelector: P1, TopologyKey: "node"}},
+	HardAntiAffinityExpressions: {{PodSelector: P2, TopologyKey: "rack"}},
+	SoftAffinityExpressions: {{PodSelector: P3, TopologyKey: "zone"}},
+	SoftAntiAffinityExpressions: {{PodSelector: P4, TopologyKey: "power"}}
+}
+```
+
+Then when scheduling pod P, the scheduler
+* Can only schedule P onto nodes that are running pods that satisfy `P1`. (Assumes all nodes have a label with key "node" and value specifying their node name.)
+* Cannot schedule P onto any racks that are running pods that satisfy `P2`. (Assumes all nodes have a label with key "rack" and value specifying their rack name.)
+* Should try to schedule P onto zones that are running pods that satisfy `P3`. (Assumes all nodes have a label with key "zone" and value specifying their zone.)
+* Should try not to schedule P onto any power domains that are running pods that satisfy `P4`. (Assumes all nodes have a label with key "power" and value specifying their power domain.)
+
+When `HardAffinityExpressions` has multiple elements, the requirements are ANDed. Likewise for `HardAntiAffinityExpressions`.
+For `SoftAffinityExpressions` and `SoftAntiAffinityExpressions`, nodes that satisfy more of the rules are preferred over those that satisfy fewer.
+
+## A comment on symmetry
+
+One thing that makes affinity and anti-affinity tricky is symmetry.
+
+Imagine a cluster that is running pods from two services, S1 and S2. Imgine that the pods of S1 have a hard anti-affinity rule
+"do not run me on nodes that are running pods from S2." It is not sufficient just to check that there are no S2 pods on a node when
+you are scheduling a S1 pod. You also need to ensure that there are no S1 pods on a node when you are scheduling a S2 pod,
+*even though the S2 pod does not have any anti-affinity rules*. Otherwise if an S1 pod schedules before an S2 pod, the S1
+pod's hard anti-affinity rule can be violated by a later-arriving S2 pod.
+
+Note that while hard anti-affinity is symmetric,
+hard affinity is *not* symmetric. That is, if the pods of S1 have a hard affinity rule "run me on nodes that are running
+pods from S2," it is not required that there be S1 pods on a node in order to schedule a S2 pod onto that node.
+
+Soft affinity and soft anti-affinity are both symmetric.
+If the pods of S1 had a soft anti-affinity rule "try not to run me on nodes that are running pods from S2"
+then we would prefer to keep a S1 pod that we are scheduling off of nodes that are running S2 pods, and also
+to keep a S2 pod that we are scheduling off of nodes that are running S1 pods. Likewise if the pods of
+S1 had a soft affinity rule "try to run me on nodes that are running pods from S2" then we would prefer
+to place a S1 pod that we are scheduling onto a node that is running a S2 pod, and also to place
+a S2 pod that we are scheduling onto a node that is running a S1 pod.
```

Comment:
Actually I think that there is another wrinkle here. If all S1 pods already have an S2 pod on their node, then subsequent S2 pods do not need to (or perhaps even should not) favour nodes that have an A pod (and implicitly also a B pod) on them already.